src/ServerRequestFactory.php

```diff
@@ -99,6 +101,52 @@ public static function fromGlobals(
         );
     }
 
+    /**
+     * Create a request from a Swoole request
+     *
+     * @param swoole_http_request $request
+     * @return ServerRequest
+     * @throws RuntimeException if Swoole is not installed
+     */
+
+    public static function fromSwoole(swoole_http_request $request)
+    {
+        if (! extension_loaded('swoole')) {
+            throw new Exception\RuntimeException('Swoole extension is not installed!');
+        }
+        $get    = isset($request->get) ? $request->get : [];
+        $post   = isset($request->post) ? $request->post : [];
+        $cookie = isset($request->cookie) ? $request->cookie : [];
+        $files  = isset($request->files) ? $request->files : [];
+
+        $server = [];
+        foreach ($request->server as $key => $value) {
+            $server[strtoupper($key)] = $value;
+        }
+
+        $headers = [];
+        foreach ($request->header as $name => $value) {
+            $headers[str_replace('-', '_', $name)] = $value;
+        }
+
+        $body = new Stream('php://temp', 'wb+');
```

Comment:
I've been thinking about this, and I'm worried this could be a performance issue and/or security vector in some cases; in particular, requests that include file uploads and/or requests with large requests bodies.

As such, I'm thinking we should create a proxy decorator class for the swoole request body that extends the current default `Stream` implementation:

- It would receive the request to the constructor.
- Any method that would normally access the underlying stream resource would:
  - check if the request body has been requested before
  - if not, it would create the `php://temp` stream and write the request body to it, and assign it via a call to `parent::construct()`.
  - otherwise, it would proxy to the parent method

In this way, we prevent duplication of the request contents _until they are accessed_, which alleviates the majority of issues (particularly since the file uploads will normally already be handled).